File: Packages/OdeteBundler/Sources/OdeteBundler/Resources/js/aquecimento.js
// Aquecimento: o pacote de dependências feito antes de alguém pedir o dev server.
//
// O primeiro `npm run dev` de um projeto recém-instalado pagava tudo de uma vez: carregar
// o esbuild, fazer o build do app e empacotar o react-dom — uns oito segundos de CPU num
// iPad sem JIT. O pacote de dependências fica em disco (dependencias.js) e vale para
// qualquer servidor com a mesma base; então ele é feito aqui, logo depois do
// `npm install` (ou quando o Preview de um projeto JS aparece), e o servidor que subir
// depois acha tudo pronto.
//
// As chaves do pacote são os arquivos de node_modules que o app importa, e quem as diz é
// o build do app com as mesmas opções do servidor (`opcoesDoBundle` do devserver.js):
// mesmas chaves, mesmo nome de pacote guardado, e o servidor o acha pelo nome exato.
(function () {
  let contador = 0;

  globalThis.__devAquece = async function (root, soSeFaltar) {
    const fs = require("fs"), path = require("path");
    const inicio = Date.now();
    let nm;
    try { nm = fs.statSync(path.join(root, "node_modules")).isDirectory(); } catch (e) { nm = false; }
    if (!nm) return { feito: false, motivo: "sem-node-modules" };
    let html;
    try { html = fs.readFileSync(path.join(root, "index.html"), "utf8"); } catch (e) { return { feito: false, motivo: "sem-index" }; }
    const entradas = [];
    html.replace(/<script\s+type="module"\s+src="([^"]+)"\s*><\/script>/g, (m, s) => {
      if (!/^[a-z][a-z0-9+.-]*:/i.test(s)) entradas.push(s.replace(/^\//, ""));
      return m;
    });
    if (!entradas.length) return { feito: false, motivo: "sem-entradas" };
    const prefixo = "aquece" + ++contador + ":";
    const deps = globalThis.__depsCria(root, prefixo);
    if (soSeFaltar && deps.jaTemPacote()) return { feito: false, motivo: "ja-tinha" };
    const pedidas = new Set();
    try {
      for (const e of entradas) {
        if (!fs.existsSync(path.join(root, e))) continue;
        // `memoria: false`: o servidor que subir depois lê tudo de novo; guardar aqui
        // seria memória presa até alguém limpar.
        const r = await globalThis.__rebuild(prefixo + e, {
          root, entries: [e], format: "esm", platform: "browser", dev: true, outdir: "__odete",
          externalMissing: true, preempacota: true, banner: 'import "/@odete/deps.js";', memoria: false,
        });
        if (r.ok && r.dependencias) for (const d of r.dependencias) pedidas.add(d);
      }
      if (!pedidas.size) return { feito: false, motivo: "sem-pacotes" };
      const g = await deps.garante([...pedidas]);
      const est = deps.estatisticas();
      return {
        feito: !!g.ok, pacotes: pedidas.size, doDisco: est.depsDoDisco > 0, ms: Date.now() - inicio,
        erro: g.ok ? null : ((g.errors && g.errors[0] && g.errors[0].text) || "erro"),
      };
    } finally {
      await globalThis.__descartaContextos(prefixo);
    }
  };
})();

File: Packages/OdeteBundler/Sources/OdeteBundler/Aquecimento.swift
import Foundation
import Synchronization

/// O esbuild carregado e o pacote de dependências feito antes do primeiro `npm run dev`.
///
/// Sem isto, o primeiro servidor de um projeto recém-instalado pagava tudo na hora em que
/// a pessoa esperava o Preview: carregar o esbuild (o módulo de 14 MB em wasm
/// interpretado), o build do app e o empacotamento do react-dom — uns oito segundos de
/// CPU num iPad sem JIT. O pacote de dependências fica em `node_modules/.odete-deps` e o
/// servidor aceita qualquer um que contenha o que ele pede; feito aqui, depois do
/// `npm install` ou quando o Preview de um projeto JS aparece, o servidor cai no caminho
/// rápido (medido: ~1 s com o pacote pronto).
///
/// Um aquecimento por projeto de cada vez. Quem vai subir o servidor espera o que estiver
/// em andamento (`esperar`): dois empacotamentos iguais no mesmo motor, ao mesmo tempo,
/// seriam o dobro do trabalho para o mesmo arquivo.
public enum Aquecimento {
    public struct Resultado: Sendable, Equatable {
        /// Fez o pacote agora (ou achou um que já servia, em `doDisco`).
        public var feito = false
        /// Por que não fez nada: `sem-node-modules`, `sem-index`, `ja-tinha`…
        public var motivo: String?
        public var pacotes = 0
        public var doDisco = false
        public var erro: String?
    }

    private static let emAndamento = Mutex<[String: Task<Resultado, Never>]>([:])

    /// Carrega o esbuild e faz o pacote de dependências de `raiz`, se der. `soSeFaltar`
    /// não faz o build do app quando já há pacote guardado para o lock de agora — é o que
    /// o Preview usa ao aparecer, para não gastar CPU à toa num projeto já aquecido.
    @discardableResult
    public static func aquecer(esbuild: Esbuild, raiz: URL, soSeFaltar: Bool = false) -> Task<Resultado, Never> {
        let chave = Esbuild.chave(raiz)
        return emAndamento.withLock { tabela in
            if let t = tabela[chave] {
                return t
            }
            let t = Task<Resultado, Never> {
                defer { _ = emAndamento.withLock { $0.removeValue(forKey: chave) } }
                return await rodar(esbuild: esbuild, raiz: raiz, soSeFaltar: soSeFaltar)
            }
            tabela[chave] = t
            return t
        }
    }

    /// Espera o aquecimento de `raiz` que estiver em andamento. Sem nenhum, volta na hora.
    public static func esperar(raiz: URL) async -> Resultado? {
        let t = emAndamento.withLock { $0[Esbuild.chave(raiz)] }
        return await t?.value
    }

    public static func emAndamento(raiz: URL) -> Bool {
        emAndamento.withLock { $0[Esbuild.chave(raiz)] != nil }
    }

    private static func rodar(esbuild: Esbuild, raiz: URL, soSeFaltar: Bool) async -> Resultado {
        do {
            _ = try await esbuild.ready()
            // Sem node_modules não há pacote para fazer: o motor carregado já é o ganho.
            guard FileManager.default.fileExists(atPath: raiz.appending(path: "node_modules").path) else {
                return Resultado(motivo: "sem-node-modules")
            }
            let js = Bundle.module.url(forResource: "js", withExtension: nil)!
            for arquivo in ["dependencias.js", "aquecimento.js"] {
                try await esbuild.engine.evaluate(
                    String(contentsOf: js.appending(path: arquivo), encoding: .utf8),
                    name: arquivo
                )
            }
            let json = try await esbuild.engine.call("__devAquece", [raiz.path, soSeFaltar])
            let obj = (try? JSONSerialization.jsonObject(with: Data(json.utf8)) as? [String: Any]) ?? [:]
            return Resultado(
                feito: (obj["feito"] as? Bool) ?? false,
                motivo: obj["motivo"] as? String,
                pacotes: (obj["pacotes"] as? Int) ?? 0,
                doDisco: (obj["doDisco"] as? Bool) ?? false,
                erro: obj["erro"] as? String
            )
        } catch {
            return Resultado(erro: error.localizedDescription)
        }
    }
}
